VMware.CloudFoundation.CertificateManagement.psm1

```diff
@@ -25,17 +25,924 @@ if ($PSEdition -eq 'Desktop') {
+        [Parameter (Mandatory = $false)] [ValidateNotNullOrEmpty()] [String] $certificateFileExt = ".cer",
+        [Parameter (Mandatory = $false)] [ValidateNotNullOrEmpty()] [String] $timeout = 18000
+    )
+
+    Try {
+        $vcfVcenterDetails = Get-vCenterServerConnection -server $server -user $user -pass $pass -domain $domain 
+        if ($PsBoundParameters.ContainsKey("cluster")) {
+            if (Get-Cluster | Where-Object { $_.Name -eq $cluster }) {
+                $esxiHosts = Get-Cluster $cluster | Get-VMHost | Sort-Object -Property Name
+                if (!$esxiHosts) { Write-Warning "No ESXi hosts found within $cluster cluster." }
+            }
+            else {
+                Write-Error "Unable to locate Cluster $cluster in $($vcfVcenterDetails.fqdn) vCenter Server: PRE_VALIDATION_FAILED" -ErrorAction Stop
+            }
+        }
+        else {
+            $esxiHosts = Get-VMHost -Name $esxiFqdn
+            if (!$esxiHosts) { Write-Error "No ESXi host '$esxiFqdn' found within workload domain '$domain'." -ErrorAction Stop }
+        }
+    
+        # Certificate replacement starts here
+        $replacedHosts = New-Object Collections.Generic.List[String]
+        $skippedHosts = New-Object Collections.Generic.List[String]
+        Foreach ($esxiHost in $esxiHosts) {
+            $esxiFqdn = $esxiHost.Name
+            $crtPath = "$certificateFolder\$esxiFqdn$certificateFileExt"
+
+            if (!(Test-Path $crtPath -PathType Leaf )) {
+                Write-Error "Could not find certificate in $crtPath. Skipping certificate replacement for $esxiFqdn. "
+                $skippedHosts.Add($esxiFqdn)
+                continue
+            }
+
+            if (Confirm-ESXiCertificateAlreadyInstalled -server $server -user $user -pass $pass -esxiFqdn $esxiFqdn -signedCertificate $crtPath) {
+                $skippedHosts.Add($esxiFqdn)
+                continue
+            }
+            else {
+                $esxiCredential = (Get-VCFCredential -resourcename $esxiFqdn | Where-Object { $_.username -eq "root" })
+                if ($esxiCredential) {
+                    
+                    Set-ESXiState -esxiFqdn $esxiFqdn -state "Maintenance" -VsanDataMigrationMode "Full" -timeout $timeout
+                
+                    Write-Host "Starting certificate replacement for $esxiFqdn"                                    
+
+                    $esxCertificatePem = Get-Content $crtPath -Raw
+                    Set-VIMachineCertificate -PemCertificate $esxCertificatePem -VMHost $esxifqdn
+                    $replacedHosts.Add($esxiFqdn)
+                    
+                    Restart-ESXiHost -esxiFqdn $esxiFqdn -user $($esxiCredential.username) -pass $($esxiCredential.password)
+                
+                    # Connect to vCenter server, then connect ESXi host to it and exit maintenance mode
+                    Write-Host "Exiting maintenance mode and connecting to vCenter"
```

Comment:
I think this should move down under $vcfVcenterDetails and then be:

`Write-Output "Connecting to vCenter Server instance $vcfVcenterDetails and exiting ESXi host $esxiFqdn from maintenance mode."`